poll/poll.py

```diff
@@ -0,0 +1,849 @@
+
+    def clean_tally(self):
+        """
+        Cleans the tally. Scoping prevents us from modifying this in the studio
+        and in the LMS the way we want to without undesirable side effects. So
+        we just clean it up on first access within the LMS, in case the studio
+        has made changes to the answers.
+        """
+        answers = OrderedDict(self.answers)
+        for key in answers.keys():
+            if key not in self.tally:
+                self.tally[key] = 0
+
+        for key in self.tally.keys():
+            if key not in answers:
+                del self.tally[key]
+
+    def tally_detail(self):
+        """
+        Return a detailed dictionary from the stored tally that the
+        Handlebars template can use.
+        """
+        tally = []
+        answers = OrderedDict(self.markdown_items(self.answers))
+        choice = self.get_choice()
+        total = 0
+        self.clean_tally()
+        source_tally = self.tally
+        any_img = self.any_image(self.answers)
+        for key, value in answers.items():
+            count = int(source_tally[key])
+            tally.append({
+                'count': count,
+                'answer': value['label'],
+                'img': value['img'],
+                'key': key,
+                'first': False,
+                'choice': False,
+                'last': False,
+                'any_img': any_img,
+            })
+            total += count
+
+        for answer in tally:
+            if answer['key'] == choice:
+                answer['choice'] = True
+            try:
+                answer['percent'] = round(answer['count'] / float(total) * 100)
+            except ZeroDivisionError:
+                answer['percent'] = 0
+
+        tally.sort(key=lambda x: x['count'], reverse=True)
+        # This should always be true, but on the off chance there are
+        # no answers...
+        if tally:
+            # Mark the first and last items to make things easier for Handlebars.
+            tally[0]['first'] = True
+            tally[-1]['last'] = True
+        return tally, total
+
+    def get_choice(self):
+        """
+        It's possible for the choice to have been removed since
+        the student answered the poll. We don't want to take away
+        the user's progress, but they should be able to vote again.
+        """
+        if self.choice and self.choice in OrderedDict(self.answers):
+            return self.choice
+        else:
+            return None
+
+    def student_view(self, context=None):
+        """
+        The primary view of the PollBlock, shown to students
+        when viewing courses.
+        """
+        if not context:
+            context = {}
+        js_template = self.resource_string(
+            '/public/handlebars/poll_results.handlebars')
+
+        choice = self.get_choice()
+
+        context.update({
+            'choice': choice,
+            # Offset so choices will always be True.
```

Comment:
@nedbat [stale comment](https://github.com/open-craft/xblock-poll/commit/70d645fd0149520abb36e4da0ac90e01d1cec9ab#diff-39725552b3964905d716e989c7ca4e23R82) - thank you for noticing, I'll remove it